### PR TITLE
More functionnalities for repl

### PR DESCRIPTION
**Pull request by OscarLahaie**
This pr is intended to be a part of the work of @ulugbekna in #677.
It would be interesting to have an evaluation system similar to Tuareg in Emacs. For this it is not only necessary to evaluate the user's selection but also when there is no selection the expression corresponding to the cursor position.
-
The objectives :
- [x]  Evaluate a file 
- [x] Preserve focus on editor after evaluations
- [x]  Evaluate expressions
- [x] Option to jump to the next expression

### CODE REVIEW

**Comment by OscarLahaie** [comment]
This is the cleaner version of pull request #943.

File: src/repl.ml
Comment (mnxn) [inline src/repl.ml:152]: This formatting sometimes looks worse like in this example with a single expression:

```
# 
  x        
  ;;       
- : int = 1
```

Maybe it's worth checking if the code is a single line first? If it is, then there's no need to add any extra formatting.

I would prefer the following in simple cases:
```
# x ;;       
- : int = 1
```

**Review by mnxn — CHANGES_REQUESTED** [review]
A few changes need to be made for Windows and the official OCaml REPL.

I haven't had a chance to try macOS or utop yet.

File: src/repl.ml
Comment (mnxn) [inline src/repl.ml:230]: `get_uri` will return paths with backslashes like `c:\a\b\c` when it uses `Uri.fsPath` on Windows. [Quoted string literals](https://v2.ocaml.org/manual/lex.html#sss:stringliterals) are more appropriate in this situation.

```suggestion
          let use = "#use {|" ^ uri ^ "|};;" in
```

File: src/repl.ml
Comment (mnxn) [inline src/repl.ml:144]: `Uri.path` returns paths like `/c:/a/b/c` on Windows, which are invalid for `#use` in the REPL. This should be `Uri.fsPath` instead.
```suggestion
  |> Vscode.Uri.fsPath
```

**Review by mnxn — CHANGES_REQUESTED** [review]

File: src/repl.ml
Comment (OscarLahaie) [reply src/repl.ml]: Done!
I have done some tests and it seems to work well with the new version of the function.
I am not sure about the `List.mem` efficiency, what do you think about this solution?

File: src/repl.ml
Comment (OscarLahaie) [reply src/repl.ml]: I have not tried it on MacOs and I can't. However, with your suggestions and the new patch, it works on Windows and Linux on `Utop` and `Ocaml`.

File: src/repl.ml
Comment (mnxn) [reply src/repl.ml]: The new behavior is great.

`String.mem` is fine but I'm not sure what the difference is from `String.contains` that we use in other parts of the extension.

**Comment by OscarLahaie** [comment]
I made several tests to see if the jump is working but I am sure some cases unknown to me.
So if you find any bugs tell me so I can fix them.

**Review by mnxn — APPROVED** [review]

**Comment by rgrinberg** [comment]
@mnxn is this good to merge?

**Comment by ulugbekna** [comment]
I will have a look at this today

**Comment by mnxn** [comment]
> is this good to merge?

I looked good to me when I tried it.

Ready to merge unless @ulugbekna finds something.

File: src/repl.ml
Comment (ulugbekna) [inline src/repl.ml:186]: I'm a little skeptical about having a 180-line function

**Review by ulugbekna — CHANGES_REQUESTED** [review]
This needs a changelog entry. 

Didn't look too closely at the 180-line function, but functionality added by the PR seems to work well

**Comment by ulugbekna** [comment]
I think this can me squashed and merged, but jumping to next expression implementation needs to be made on the ocamllsp side as a new custom request so that we avoid hacks that we have now.

**Comment by OscarLahaie** [comment]
> I think this can me squashed and merged, but jumping to next expression implementation needs to be made on the ocamllsp side as a new custom request so that we avoid hacks that we have now.
> 
> Also, one should be careful using `registerTextEditorCommand` because it's supposed to be run synchronously, meaning starting off a promise as we do it now is unsafe. See [microsoft/vscode#16814](https://github.com/microsoft/vscode/issues/16814)

I don't have much time this week. Nevertheless I will look at and improve the code in the coming weeks if necessary. Because there are some behaviours to be fixed at the ocaml-lsp level.

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:134]: Why do you obtain `document` right away but only use it in one branch of the conditional?

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:150]: As far as I can tell `String.mem trimmed_code '\n'` is always evaluated. Therefore it should be assigned a binding and moved before the conditionals.

File: README.md
Comment (rgrinberg) [inline README.md:267]: Why don't we preserve `ocaml.evaluate-selection`? I don't see anything wrong with having both commands.

**Review by rgrinberg — CHANGES_REQUESTED** [review]

**Comment by rgrinberg** [comment]
Before merging this, I think we need to get on the same page of what good REPL functionality looks like. I have no interest in copying tuareg and its limitations. We can do a lot better if we think about things a bit more.

File: README.md
Comment (rgrinberg) [reply README.md:267]: I understood. I just don't see why we should remove a command with a clear name and semantics and replace it with a command that has neither of those things. Seems like we're just irritating users who are used to the old command this way.

**Comment by ulugbekna** [comment]
> Before merging this, I think we need to get on the same page of what good REPL functionality looks like. I have no interest in copying tuareg and its limitations. We can do a lot better if we think about things a bit more.

Do you have precise limitations you'd like gone before merging? 

Things that are not addressed in this PR but can be done in other, IMO: 

- configuration for how module is evaluated (#use or #mod_use) as a dialog 
- configurable REPL instance -- the best decision is to have two commands for picking a terminal to send commands to:
  - one command to pick an integrated terminal that already exists from a list (this allows users to jump into existing repl session and send commands to it, for example) 
  - one command to launch a terminal from settings-defined repl binary and args OR default utop - this already exists but it uses a timeout before sending commands to a new REPL -- this needs to be fixed

File: README.md
Comment (ulugbekna) [reply README.md:267]: Good point, but then I would show a deprecation message saying to use the new command because 

1. The new command has almost the same semantics - select and evaluate (no selection resulted in evaluating the whole line but I'm not sure that's very useful and the new command does the same thing if it's a single-line let binding) 
2. There's then contention for the shortcut: keep the old and miss out on a better command or replace and still irritate

**Comment by rgrinberg** [comment]
First of all, as a general comment, there's just too many knobs, settings, commands that all make the experience way too complicated. I've already went through this stuff with tuareg mode. In the end, their toplevel integration is hardly usable.

99% of the time when I use the toplevel, I do the following:

1. Open a project
2. Navigate to a module I want to experiment with
3. Figure out how to load into the toplevel.

We should focus on this core feature first before we worry about directives, selecting repl instances. Finally, we should absolutely avoid any code that tries to parse OCaml. I see this PR is adding quite a bit of that and it's really wasted effort.

Here's a simple idea for how to make the toplevel more useful.

* We add a command `$ dune ocaml top-module path/to/module.ml --source=/tmp/selection/file`. This command will either load the toplevel binary or print the directives required to load `module.ml` as if its source was `--source`.

* We add a request to lsp that given a position in the document, will return a completed & pretty printed parse tree. Luckily for us, merlin already knows how to do this.

* On the vscode side, we add a command to launch a toplevel where the cursor is. It will obtain the source using lsp, then run the dune command mentioned above. There's no need to select toplevel instances. We associate every module with a toplevel instance. If there's already a toplevel for a module, then we kill it and launch it again.

Let's forget about all the other stuff for now.

File: src/repl.ml
Comment (rgrinberg) [inline src/repl.ml:308]: When can this request return None?

**Comment by smorimoto** [comment]
Close this for now, but feel free to re-open.

**Pull request closed without merging**
